sw/src/offline-interface/offline-interface.js

```diff
@@ -0,0 +1,191 @@
+import EventEmitter from 'events'
+import i18n from '@dhis2/d2-i18n'
+import { swMsgs } from '../lib/constants'
+import { register, unregister, checkForUpdates } from '../lib/registration'
+import { openSectionsDB, SECTIONS_STORE } from '../lib/sections-db'
+
+/** Helper to simplify SW message sending */
+function swMessage(type, payload) {
+    if (!navigator.serviceWorker.controller)
+        throw new Error(
+            '[Offine interface] Cannot send service worker message - no service worker is registered'
+        )
+    navigator.serviceWorker.controller.postMessage({ type, payload })
+}
+
+/**
+ * Provides an interface between the client and the service worker.
+ * The interface sends and receives messages from the service worker for
+ * the purposes of 'cacheable section' recording and returns some functions
+ * that interact with the indexedDB and cache storage APIs.
+ */
+export class OfflineInterface {
+    constructor() {
+        this.pwaEnabled = process.env.REACT_APP_DHIS2_APP_PWA_ENABLED === 'true'
+
+        if (this.pwaEnabled) {
+            register()
+        } else {
+            unregister()
+        }
+
+        if ('serviceWorker' in navigator)
+            navigator.serviceWorker.oncontrollerchange = () =>
+                window.location.reload()
```

Comment:
SW registration happens upon instantiation (which happens in `/adapter/src/index.js`)